### ISSUE TITLE
Abstract method with same pre and post behavior

### ISSUE THREAD

**Issue by RezwanArefin01** [nus-cs2103-AY2223S1/forum#8]
Suppose you have an abstract method. Now every class that implements that method have a common beginning and common ending code. How would you reduce code duplication? My idea is the following:
```java
abstract class Test {

  private void preMethod() {} 
  private void postMethod() {}
  
  public void callThisMethod() { 
    preMethod(); 
    method();
    postMethod(); 
  }
  
  public abstract method(); 
}
```
However this approach doesn't prevent anyone from calling `method()` directly, which will be a problem. Making it `private` isn't an option, since abstract method cannot be `private`.
With this approach, a proper access modifier would be one that only sub-classes can access and inherit, but other package classes cannot. But I don't think any of the access modifier does that. So my current option is to ensure that I don't call `method()` directly from other classes myself.

Any thoughts on how to improve this?

**Comment by RichDom2185** [comment]
I think it depends on what you are trying to do with `method`. Does it need to access attributes of the superclass? Maybe you can try using `CompletableFuture`. Or you could use a functional interface and pass in the method as an argument to something. Or maybe expose the `preMethod` and `postMethod` to subclasses and call them from there. I think without an example use-case, it would be hard to decide on a most appropriate implementation because it depends on what the end goal is...

**Comment by RichDom2185** [comment]
Either way, your proposed behaviour seems to be a good candidate for implementing the decorator design pattern. You can learn more about it [here](https://refactoring.guru/design-patterns/decorator/java/example).

```java
interface Methodable {
  void method();
}

class TestDecorator extends Methodable {
  private final Methodable wrapee;

  TestDecorator(Methodable wrapee) {
    this.wrapee = wrapee;
  }

  private void preMethod() {} 
  private void postMethod() {}

  @Override
  public void method() {
    preMethod();
    wrapee.method();
    postMethod();
  }
}
```

_PS: The website has lots of cool stuff!_

**Comment by RezwanArefin01** [comment]
@RichDom2185 
> expose the preMethod and postMethod to subclasses and call them from there.   

This is the code duplication I am trying to avoid. Also I would not prefer to involve functional programming. 

The use case for this is: `method()` runs a command with arguments, and the `preMethod()` checks whether the command and the arguments are compatible. What I am trying to achieve is: 
```java 
callThisMethod(args) { 
  if (!preMethod(args)) return; 
  method(args);
}
```
(This isn't actually what I wrote in the post though; because of the `return`).

**Comment by RichDom2185** [comment]
Hmm it seems like you are trying to add some sort of "input validation" functionality. I would suggest using the decorators, which is an object-oriented design pattern.

**Comment by RezwanArefin01** [comment]
Yes, it is input validation but I want to do this in one place only. 
I'll have to read more about decorators. On first look, it doesn't seem very appealing...

**Comment by deeyonn** [comment]
I'm not sure if I understood the requirements for method(), but why wouldn't a simple producer class fit the role for method()?

https://docs.oracle.com/javaee/7/api/javax/enterprise/inject/spi/Producer.html

https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html

**Comment by RezwanArefin01** [comment]
@deeyonn I think you linked the wrong `Producer` documentation page (`Producer` was cs2030s specific, it is actually called `Supplier` in Java), but yes it would work. I think this approach is identical to what @RichDom2185 suggested. We'd need to wrap the `method()` in an object (`Methodable` / `Producer` / `Runnable` / `Consumer` etc) and the sub-classes would need to provide that object. 
However, how would you enforce the sub-classes to provide the wrapper object? For methods it would be checked by compiler, since sub-classes must provide implementation of abstract methods.

**Comment by Eclipse-Dominator** [comment]
Hello, regarding the original post 

> However this approach doesn't prevent anyone from calling `method()` directly, which will be a problem. Making it `private` isn't an option, since abstract method cannot be `private`. With this approach, a proper access modifier would be one that only sub-classes can access and inherit, but other package classes cannot. But I don't think any of the access modifier does that. So my current option is to ensure that I don't call `method()` directly from other classes myself.
> 
> Any thoughts on how to improve this?

I think you can always just use `protected` access modifier to achieve that level of access you have mentioned

**Comment by RezwanArefin01** [comment]
@Eclipse-Dominator `protected` is not ideal for this case. It doesn't prevent other classes of the same package calling the function. We'd still need to ensure that we don't call the function ourselves, compiler won't check!
